### DISCUSSION TITLE
Features you'd like to see here

### DISCUSSION THREAD

**Discussion by santoshlite** [ssantoshp/Empyrial#13] [Ideas]
Hey,

We are curious to know what are features you'd like to see in Empyrial?

Shoot! 🔫

**Comment by agn35** [comment]
Great work! Thank you! For the portfolio optimizer, it would be helpful if the font of stocks in the pie chart could be changed. Similarly, 0% stocks should not be listed in the portfolio optimization pie chart. Last, I wonder whether it is normal that an unoptimized portfolio performs better than the optimized portfolio...

**Reply by santoshlite** [reply]
Hey Adam,

You can do it by doing that: https://colab.research.google.com/drive/143-BJTZe9eH52yvgG5KU7PhSBp0FXrqq?usp=sharing

Disclaimer: the first time you'll do it, it may give you an error (about Datetimeindex), in that case just restart and run it all again (you have this option on colab and on Jupyter notebook)

You can also notice that the 0% has been removed from the chart :)

Yes, it's normal, portfolio optimization is not always better (but most of the time it is), it depends on your assets (it's not a magic formula that works all the time unfortunately 😞 , you can try different types of optimizations and see which one works best with your portfolio.

**Comment by agn35** [comment]
Hi Santosh,

Thank you for getting back to me. I can easily run the script (see output below), but this prompts a few more questions: 
1) The font size seems to change heading and legend inversely; when font size goes up, heading gets smaller. When font size decreases, heading gets bigger and legend smaller. Probably good to add a legend_font command. 
2) How can I control the color of the various pie pieces? I would want to avoid black altogether and also don’t want two colors that are too close to one another (such as BLK and NVDA). 
3) Running the script on a diworsified portfolio (see second screenshot), the script still includes 0 percentages for all the stocks it doesn’t want. 

Best,
Adam

**Reply by santoshlite** [reply]
Hey,

We are working on the software part and we will solve that there. I'll come back to you very soon.

Best,

**Reply by santoshlite** [reply]
Hi,

Sorry for the late reply. 1) and 2) is fixed in the new version, you can try it. 
For 3 you can do the same as in this notebook: https://colab.research.google.com/drive/1pKzzJjySy6_wbI8TKDKpHX1ZLXlTNCOZ?usp=sharing

Hope it helps!

**Comment by uberscientist** [comment]
I'm looking forward to using Empyrial more, I'd like to have the ability to test rebalancing rules (bi-annually, quarterly, monthly, etc.)
Also curious about exploring volatility triggered rebalancing, or allocation % drift based rebalancing rules

Thanks for the open source goodies :)

**Reply by santoshlite** [reply]
Hey there! A lot of people asked for the rebalancing features and it is currently in development, it will be finish very soon (in less than a week), we'll keep you posted 😁

**Reply by santoshlite** [reply]
And also with that will come an order book!

**Reply by santoshlite** [reply]
We just added the rebalancing feature you were talking about (for the moment it's calendar-based only). You can check the release [here](https://github.com/ssantoshp/Empyrial/releases/tag/1.5.0)!

An example of how to use it [here](https://github.com/ssantoshp/Empyrial/blob/main/Empyrial/test/rebalance_test.py)

Let me know how it goes :))

**Reply by uberscientist** [reply]
Great, just ran the example, worked well and checked the code for how to format the arg:  `acceptable_schedules = ["6mo","1y","2y", "quarterly", "monthly"]`

Thanks again, very cool work

*EDIT*: Just saw the readme was updated with the new rebalance feature too, bravo

**Reply by santoshlite** [reply]
Nice, thanks! ✌️

**Reply by gurusura** [reply]
It would be great if you also have a simple import of holdings from various brokerages. May be also include nsepy or nsetools instead of just yfinance if it makes sense to you.
https://nsetools.readthedocs.io/en/latest/usage.html
https://nsepy.xyz/

**Reply by santoshlite** [reply]
Good idea I actually considered adding brokerage supports. It'll certainly be later in our roadmap in our software. Including nsepy do makes sense indeed, we'll consider adding that to the project too, thanks :)

**Reply by santoshlite** [reply]
@gurusura It's been a long time but here we are: [v2.1.3](https://github.com/ssantoshp/Empyrial/releases/tag/v2.1.3) supports custom data. Checkout out the doc!

**Comment by mbatzeris** [comment]
B R I N G  B A C K  T H E   O R A C L E ! ! !

**Reply by santoshlite** [reply]
I'll try that 👍

**Reply by santoshlite** [reply]
@mbatzeris Here it comes: https://github.com/ssantoshp/The-Oracle

**Comment by pbajpayee** [comment]
Hi Santosh
Is it possible to run a portfolio backtest with variable asset? For example, monthly rebalancing allows you to replace existing tickers with new ones?

**Reply by santoshlite** [reply]
Hey, no it isn't sorry!

**Comment by Jeraz** [comment]
Hi Santosh,

Is there a way to integrate this with live trading platform for portfolio management?

Thanks!

**Reply by santoshlite** [reply]
Which platform are you thinkig about?

**Reply by Jeraz** [reply]
I am thinking of Robinhood or Alpaca as they have API support.

**Reply by santoshlite** [reply]
So you would empyrial to be able to connect to your brokerage account to do backtesting?

**Reply by Jeraz** [reply]
Yes I should be able to do that. Alpaca offers backtesting through paper money.

**Reply by santoshlite** [reply]
You could do it if there is a way to export your price data; that involves some manual work. But connecting to Alpaca sounds unlikely to happen immediately since that's quite a specific feature. I am not fully sure what is the motivation to do so since Alpaca already offers backtesting. Do you want more performance and risk metrics? I am not familiar with what alpaca offers...

**Comment by piper0124** [comment]
Hi Santosh：
Can I replace the default benchmark （"SPY"）with custom data，or use my own data as the benchmark？
THANKS :)
